imap_processing/ultra/l1b/ultra_l1b.py

```diff
@@ -15,67 +20,92 @@ def create_dataset(data_dict, name):
-    # For now we are using it for retrieving the epoch.
-    dataset = data_dict["imap_ultra_l1a_45sensor-de"]
-
-    # Load metadata from the metadata file
-    with open(Path(__file__).parent.parent / "ultra_metadata_example.json") as f:
-        metadata = json.loads(f.read())[name]
-
-    epoch = dataset.coords["epoch"]
-
-    annotated_de_attrs = metadata["dataset_attrs"]
+    cdf_manager = CdfAttributeManager(Path(__file__).parents[2] / "cdf" / "config")
+    cdf_manager.load_global_attributes("imap_default_global_cdf_attrs.yaml")
+    cdf_manager.load_global_attributes("imap_ultra_global_cdf_attrs.yaml")
+    cdf_manager.load_variable_attributes("imap_ultra_l1b_variable_attrs.yaml")
+
+    epoch_time = xr.DataArray(
+        data_dict["epoch"],
+        name="epoch",
+        dims=["epoch"],
+        attrs=cdf_manager.variable_attributes["epoch"],
+    )
 
     dataset = xr.Dataset(
-        coords={"epoch": epoch},
-        attrs=annotated_de_attrs,
+        coords={"epoch": epoch_time},
+        attrs=cdf_manager.get_global_attributes(name),
     )
 
-    dataset["x_front"] = xr.DataArray(
-        np.zeros(len(epoch), dtype=np.float32),
-        dims=["epoch"],
-        attrs=metadata["x_front"],
-    )
+    for key in data_dict.keys():
+        if key == "epoch":
+            continue
+        dataset[key] = xr.DataArray(
+            data_dict[key],
+            dims=["epoch"],
+            attrs=cdf_manager.variable_attributes[key],
+        )
 
     return dataset
 
 
-def ultra_l1b(data_dict: dict):
+def ultra_l1b(dependencies: list[Path]):
     """
     Process ULTRA L1A data into L1B CDF files at output_filepath.
 
     Parameters
     ----------
-    data_dict : dict
-        The data itself and its dependent data.
+    dependencies : list[Path]
+        List of dependencies.
 
     Returns
     -------
     output_datasets : list of xarray.Dataset
         List of xarray.Dataset
     """
+    data_dict = {}
+    for dependency in dependencies:
+        dataset = load_cdf(dependency, to_datetime=True)
+        data_dict[dataset.attrs["Logical_source"]] = dataset
+
     output_datasets = []
     instrument_id = 45 if "45" in next(iter(data_dict.keys())) else 90
 
-    # TODO: Add the other l1b products here.
-    l1b_products = [
-        f"imap_ultra_l1b_{instrument_id}sensor-de",
-        # f"imap_ultra_l1b_{instrument_id}extended-spin",
-        # f"imap_ultra_l1b_{instrument_id}culling-mask",
-        # f"imap_ultra_l1b_{instrument_id}badtimes"
-    ]
-
-    for name in l1b_products:
-        # TODO: perform l1b calculations here.
-        # For now we are using the L1A data (data_dict) as a placeholder.
-        dataset = create_dataset(data_dict, name)
+    if f"imap_ultra_l1a_{instrument_id}sensor-rates" in data_dict:
+        extendedspin_dict = calculate_extendedspin(data_dict)
+        extendedspin_dataset = create_dataset(
+            extendedspin_dict, f"imap_ultra_l1b_{instrument_id}sensor-extendedspin"
+        )
```

Comment:
I see now, you're calling your create_dataset function here from those dictionaries. Can you pull this `create_dataset` call into the `calculate_extended_spin` function so you are going dataset -> dataset in that routine?